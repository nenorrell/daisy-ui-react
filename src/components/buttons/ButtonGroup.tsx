import {forwardRef, HTMLAttributes, PropsWithChildren } from "react";
import clsx from "clsx";

/** For responsivene direction, please leverage className */
export interface IButtonGroup extends PropsWithChildren<HTMLAttributes<HTMLDivElement>>{
    isVertical ?:boolean
}
export const ButtonGroup = forwardRef<HTMLDivElement, IButtonGroup>((
    {
        isVertical,
        className,
        children,
        ...props
    },
    ref
)=>{
    return (
        <div {...props} ref={ref} className={clsx(
            "btn-group",
            isVertical && "btn-group-vertical",
            className
        )}>
            {children}
        </div>
    );
});

ButtonGroup.displayName = "ButtonGroup";
