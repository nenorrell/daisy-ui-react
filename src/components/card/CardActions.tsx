import { forwardRef, HTMLAttributes, PropsWithChildren } from "react";
import clsx from "clsx";

interface ICardActions extends PropsWithChildren<HTMLAttributes<HTMLDivElement>>{}
export const CardActions = forwardRef<HTMLDivElement, ICardActions>((
    {
        className,
        children,
        ...props
    },
    ref
)=>{
    return (
        <div {...props} ref={ref} className={clsx(
            "card-actions",
            className
        )}>
            {children}
        </div>
    );
});

CardActions.displayName = "CardActions";
