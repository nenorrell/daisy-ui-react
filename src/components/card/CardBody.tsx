import { FC, ForwardedRef, forwardRef, PropsWithChildren } from "react";
import clsx from "clsx";

interface ICardBody {
    id ?:string
    className ?:string
}
export const CardBody :FC<PropsWithChildren<ICardBody>> = forwardRef((
    {id, className, children},
    ref ?:ForwardedRef<HTMLDivElement>
)=>{
    return (
        <div ref={ref} id={id} className={clsx(
            "card-body",
            className
        )}>
            {children}
        </div>
    );
});

CardBody.displayName = "CardBody";
