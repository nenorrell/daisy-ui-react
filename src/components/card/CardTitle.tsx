import clsx from "clsx";
import { FC, ForwardedRef, forwardRef, PropsWithChildren } from "react";

interface ICardTitle {
    id ?:string
    className ?:string
}
export const CardTitle :FC<PropsWithChildren<ICardTitle>> = forwardRef((
    {id, className, children},
    ref ?:ForwardedRef<HTMLDivElement>
)=>{
    return (
        <div id={id} ref={ref} className={clsx(
            "card-title",
            className
        )}>
            {children}
        </div>
    );
});
