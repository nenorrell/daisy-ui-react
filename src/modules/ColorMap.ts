import { Variant, VariantColors } from "../@types/Daisy";


interface ColorMap<K, V> extends Map<K, V> {
    get(key: K): V;
}

// Cast to ColorMap interface to get rid of inherit "possibly undefined value" from Map
export default <ColorMap<Variant, VariantColors>>new Map([
    ["primary", {
        bg: "bg-primary",
        text: "text-primary-content",
        focus: "primary-focus",
        btn: "btn-primary"
    }
    ],
    ["secondary", {
        bg: "bg-secondary",
        text: "text-secondary-content",
        focus: "secondary-focus",
        btn: "btn-secondary"
    }
    ],
    ["accent", {
        bg: "bg-accent",
        text: "text-accent-content",
        focus: "accent-focus",
        btn: "btn-accent"
    }
    ],
    ["neutral", {
        bg: "bg-neutral",
        text: "text-neutral-content",
        focus: "neutral-focus",
        btn: "btn-neutral"
    }
    ],
    ["info", {
        bg: "bg-info",
        text: "text-info-content",
        focus: "info-focus",
        btn: "btn-info"
    }
    ],
    ["success", {
        bg: "bg-success",
        text: "text-success-content",
        focus: "success-focus",
        btn: "btn-success"
    }
    ],
    ["error", {
        bg: "bg-error",
        text: "text-error-content",
        focus: "error-focus",
        btn: "btn-error"
    }
    ],
    ["warning", {
        bg: "bg-warning",
        text: "text-warning-content",
        focus: "warning-focus",
        btn: "btn-warning"
    }
    ],
    ["base-100", { bg: "bg-base-100", text: "text-base-content" }],
    ["base-200", { bg: "bg-base-200", text: "text-base-content" }],
    ["base-300", { bg: "bg-base-300", text: "text-base-content" }]
]);
