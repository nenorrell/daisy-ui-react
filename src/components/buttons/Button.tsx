import { FC, ForwardedRef, forwardRef, MouseEventHandler, PropsWithChildren } from "react";
import { BaselessVariant, ButtonSize } from "../../@types/Daisy";
import ColorMap from "../../modules/ColorMap";
import { getButtonSize } from "../../modules/utility";
import clsx from "clsx";

interface IButton {
    buttonText ?:string
    size ?:ButtonSize
    className ?:string
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler
    isGhost ?:boolean
    isLink ?:boolean
    isOutline ?:boolean
    isDisabled ?:boolean
    isGlass ?:boolean
    isLoading ?:boolean
    noAnimation ?:boolean
    shape ?:"circle" | "square"
}
export const Button :FC<PropsWithChildren<IButton>> = forwardRef((
    props,
    ref ?:ForwardedRef<HTMLButtonElement>
)=>{
    const variantColors = ColorMap.get(props.variant || "neutral");
    const btnSize = getButtonSize(props.size);

    return (
        <button
            ref={ref}
            onClick={props.onClick || undefined}
            className={
                `btn ${btnSize} ${variantColors?.btn} ${clsx(
                    props.className && props.className,
                    props.isGhost && "btn-ghost",
                    props.isGlass && "glass",
                    props.isLink && "btn-link",
                    props.isLoading && "loading",
                    props.isOutline && "btn-outline",
                    props.noAnimation && "no-animation",
                    props.shape &&
                        props.shape === "circle" && "btn-circle" ||
                        props.shape === "square" && "btn-square"
                )}`
            }
            disabled={props.isDisabled}
        >
            {props.buttonText || props.children}
        </button>
    );
});
