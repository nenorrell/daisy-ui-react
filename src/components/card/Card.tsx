import { FC, ForwardedRef, forwardRef, PropsWithChildren } from "react";
import { Variant } from "../../@types/Colors";
import clsx from "clsx";
import { twMerge } from "tailwind-merge";
import { getTextColor } from "../../modules/colors";

interface ICard {
    id ?:string
    className ?:string
    variant ?:Variant
    isBordered ?:boolean
    imageFull ?:boolean
    size ?: "normal" | "compact"
    isSide ?:boolean
}
export const Card :FC<PropsWithChildren<ICard>> = forwardRef((
    {variant="neutral", size="normal", ...props},
    ref ?:ForwardedRef<HTMLDivElement>
)=>{
    return (
        <div id={props.id} ref={ref} className={twMerge(clsx(
            "card",
            {
                [`bg-${variant}`]: variant,
                [`text-${getTextColor(variant)}-content`]: variant
            },
            size === "compact" ? "card-compact" : "card-normal",
            props.isSide && "card-side",
            props.imageFull && "image-full",
            props.className
        ))}>
            {props.children}
        </div>
    );
});

Card.displayName = "Card";
