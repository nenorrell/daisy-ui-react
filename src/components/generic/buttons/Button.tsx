import { MouseEventHandler } from "react";
import { BaselessVariant, ButtonSize } from "../../../@types/Daisy";
import ColorMap from "../../../modules/ColorMap";
import { getButtonSize } from "../../../modules/utility";

interface IButton {
    buttonText ?:string
    size ?:ButtonSize
    /** Use className for modifiers like btn-ghost, btn-circle, etc */
    className ?:string
    variant ?:BaselessVariant
    onClick ?:MouseEventHandler
}
export const Button = ({className, variant="neutral", size, buttonText, onClick, children}:React.PropsWithChildren<IButton>)=>{
    const variantColors = ColorMap.get(variant);
    const btnSize = getButtonSize(size);

    return (
        <button onClick={onClick ? onClick : undefined} className={`btn ${btnSize} ${variantColors?.btn} ${className ? className : ""}`}>
            {buttonText || children}
        </button>
    );
};
