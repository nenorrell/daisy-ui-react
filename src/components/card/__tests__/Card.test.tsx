import { render } from "@testing-library/react";
import ColorMap from "../../../modules/ColorMap";
import { VariantOptions } from "../../../modules/testUtils";
import { Button } from "../../buttons/Button";
import { Card } from "../Card";
import { CardActions } from "../CardActions";
import { CardBody } from "../CardBody";
import { CardTitle } from "../CardTitle";

describe("Card component", () => {
    describe("Variants", ()=>{
        VariantOptions.forEach(variant=>{
            it(`Respects ${variant} variant`, ()=>{
                const component = render(
                    <Card variant={variant}>
                        <CardBody>
                            <CardTitle><h2>Testing variant {variant}</h2></CardTitle>
                            <p>Test body</p>
                        </CardBody>
                    </Card>
                );
                const container = component.container.querySelector("div.card");
                const colors = ColorMap.get(variant);

                expect(container).toHaveClass(colors.bg, colors.text);
                expect(container).toMatchSnapshot();
            });
        });
    });

    it("renders with default props", () => {
        const { container } = render(
            <Card>
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.firstChild;
        expect(card).toBeInTheDocument();
        expect(card).toHaveClass("card card-normal");
    });

    it("renders with custom className", () => {
        const { container } = render(
            <Card className="find-me">
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.firstChild;
        expect(card).toBeInTheDocument();
        expect(card).toHaveClass("find-me");
    });

    it("renders with an id", () => {
        const { container } = render(
            <Card id="find-me">
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.querySelector("div#find-me");
        expect(card).toBeInTheDocument();
    });

    it("renders with size compact", () => {
        const { container } = render(
            <Card size="compact">
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.querySelector("div.card");
        expect(card).toHaveClass("card-compact");
    });

    it("renders isSide prop", () => {
        const { container } = render(
            <Card isSide>
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.querySelector("div.card");
        expect(card).toHaveClass("card-side");
    });

    it("renders with imageFull prop", () => {
        const { container } = render(
            <Card imageFull>
                <figure><img src="https://placeimg.com/400/225/arch" alt="Shoes" /></figure>
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
            </Card>
        );
        const card = container.querySelector("div.card");
        expect(card).toHaveClass("image-full");
    });

    it("renders full card", () => {
        const { container } = render(
            <Card>
                <figure className="find-me"><img src="https://placeimg.com/400/225/arch" alt="Shoes" /></figure>
                <CardBody>
                    <CardTitle><h2>Test Card</h2></CardTitle>
                    <p>Test body</p>
                </CardBody>
                <CardActions>
                    <Button>Test Button</Button>
                </CardActions>
            </Card>
        );
        const card = container.querySelector("div.card");
        const cardFigure = card?.querySelector("figure.find-me");
        const cardBody = card?.querySelector("div.card-body");
        const cardTitle = cardBody?.querySelector("div.card-title");
        const cardActions = card?.querySelector("div.card-actions");

        expect(card).toBeInTheDocument();
        expect(cardFigure).toBeInTheDocument();
        expect(cardBody).toBeInTheDocument();
        expect(cardTitle).toBeInTheDocument();
        expect(cardActions).toBeInTheDocument();
    });
});
