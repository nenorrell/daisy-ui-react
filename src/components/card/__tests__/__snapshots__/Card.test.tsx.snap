// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Variants Respects accent variant 1`] = `
<div
  class="card bg-accent text-accent-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        accent
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects base-100 variant 1`] = `
<div
  class="card bg-base-100 text-base-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        base-100
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects base-200 variant 1`] = `
<div
  class="card bg-base-200 text-base-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        base-200
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects base-300 variant 1`] = `
<div
  class="card bg-base-300 text-base-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        base-300
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects error variant 1`] = `
<div
  class="card bg-error text-error-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        error
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects error variant 2`] = `
<div
  class="card bg-error text-error-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        error
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects info variant 1`] = `
<div
  class="card bg-info text-info-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        info
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects info variant 2`] = `
<div
  class="card bg-info text-info-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        info
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects neutral variant 1`] = `
<div
  class="card bg-neutral text-neutral-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        neutral
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects primary variant 1`] = `
<div
  class="card bg-primary text-primary-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        primary
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects secondary variant 1`] = `
<div
  class="card bg-secondary text-secondary-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        secondary
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects success variant 1`] = `
<div
  class="card bg-success text-success-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        success
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects success variant 2`] = `
<div
  class="card bg-success text-success-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        success
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects warning variant 1`] = `
<div
  class="card bg-warning text-warning-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        warning
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;

exports[`Card component Variants Respects warning variant 2`] = `
<div
  class="card bg-warning text-warning-content card-normal"
>
  <div
    class="card-body"
  >
    <div
      class="card-title"
    >
      <h2>
        Testing variant 
        warning
      </h2>
    </div>
    <p>
      Test body
    </p>
  </div>
</div>
`;
